<a href="https://colab.research.google.com/github/anyoneabd2044-art/machine-learning-from-basics/blob/main/python_10.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# Variable and Memory References

In [ ]:
# value in variable-->run --> physically --> device--> ram--> memory allocate hota hai

In [ ]:
#  when our program is at rest --> it is in --> Rom(hard disk)

In [ ]:
# when our program start to run -->it load in --> Ram--> and started allocating linebyline

*Every Thing Is Inside Ram*

**As Ram is Volatile---> IF we off the program / machine , the memory occupied is removed**

*Where does everything is Stored In Ram*

**It Is Stored In REGiSTERS**

#
#

*REGISTER are tiny storage unit inside CPU*

*RAM is a seperate memory system made of capacitors or flip-flops on memory chips*

**CPU reads/write data from RAM into registers when performing operations**

#
#

**CPU always tries to load data from RAM into register for Computation**

**Register is a very small, very fast memory storage built inside the CPU**

In [ ]:
# When We Create a "Variable"--> Pyhton called it "Name"

In [ ]:
a

'HELLOworld'

In [ ]:
a = 4

In [ ]:
id(a)

11654472

In [ ]:
hex(11642184)

'0xb1a548'

In [ ]:
id(4)

11654472

In [ ]:
# When We Are Creating a "Name"--> It Is Actually "Pointing To The Address" of That Particular Thing

**Call By Object Reference**

*Call by Object Reference--> Pointing toward the value/address*

#
#


# Aliasing

In [ ]:
a = 5

In [ ]:
b = a

In [ ]:
c = b

In [ ]:
id(a)

11654504

In [ ]:
id(b)

11654504

In [ ]:
id(c)

11654504

In [ ]:
del a

In [ ]:
id(a)

NameError: name 'a' is not defined

In [ ]:
b

5

In [ ]:
del b

In [ ]:
id(b)

NameError: name 'b' is not defined

In [ ]:
id(c)

11654504

In [ ]:
a=5

In [ ]:
b=a

In [ ]:
a=6

In [ ]:
b

5

# Reference Counting

In [ ]:
import sys

In [ ]:
a ="corona"

In [ ]:
b = a

In [ ]:
c = b

In [ ]:
id(a)

135606217334944

In [ ]:
id(b)

135606217334944

In [ ]:
id(c)

135606217334944

In [ ]:
sys.getrefcount(a)

4

# Garbage Collection

**Garbage Collection: is a process by which python automatically frees up memory by removing objects that are no longer in use**

*When we create variables, lists, dictionaries, etc. python stores them in memory(RAM). If you don't need them anymore, but they still stays in memory, your program will waste mempry.*

**Garbage Collector: python's gc module control the garbage collection process. It helps manage cyclic references-situations where two or more objects refers to each other**

# Weird Stuff

**WB1**

In [ ]:
a = 2

In [ ]:
b = a

In [ ]:
c = b

In [ ]:
sys.getrefcount(a)

4294967295

*Reason for weird value: The integer 2 is an interned object used all over Python; its reference count can overflow or brhave unexpectedly*

*The huge number(4294967295):This is an overflowed 32 bit unsigned integer value-not an actual reference count*

**WB2**

In [ ]:
a = 4

In [ ]:
b = 4

In [ ]:
sys.getrefcount(a)

4294967295

In [ ]:
sys.getrefcount(b)

4294967295

*It is Python reference count overflow issue.*

*This happenes because: .Small integer like 0-256 are interned (Python keeps a a single shared copy of these values in memory for eficiency)*

*These small integer are used everywhere in Python-inside loops, constant, function calls, etc.*

**So their reference count become extremely large**

**The Real Reference Count is too large to fit in a 32-bit integer-it overflowed**

In [ ]:
a  = 256

In [ ]:
b = 256

In [ ]:
id(a)

11662536

In [ ]:
id(b)

11662536

In [ ]:
a = 257

In [ ]:
b = 257

In [ ]:
id(a)

135606217855536

In [ ]:
id(b)

135606217856112

**This is "integer caching" or "interning"**

*This happenes beause: Python automatically caches(or interns)small integer objects in memory to save space and improve performance*

*The range of cached integer is [-5,256]*

**Integer between -5 and 256 are pre-created and reused by Python.**

**Integer outside this range are created as new object each time**

In [ ]:
a = "haldia"

In [ ]:
b ="haldia"

In [ ]:
id(a)

135606217339216

In [ ]:
id(b)

135606217339216

In [ ]:
a = "haldia inst tech"

In [ ]:
b = "haldia inst tech"

In [ ]:
id(a)

135606218035312

In [ ]:
id(b)

135606218035632

In [ ]:
a = "haldia_inst_tech"

In [ ]:
b = "haldia_inst_tech"

*This is not a valid identifier*

In [ ]:
id(a)

135606218036784

In [ ]:
id(b)

135606218036784

**This will happens when there is a "Valid identifier"**

*valid identifier: starts with alph or have underscore*

In [ ]:
L=[1,2,3]

In [ ]:
id(L)

135606218036352

In [ ]:
id(L[0])

11654376

In [ ]:
id(1)

11654376

In [ ]:
id(2)

11654408

In [ ]:
id(3)

11654440

In [ ]:
L[2]=1

In [ ]:
id(L[2])

11654376

In [ ]:
L=[1,2,3,[4,5]]

In [ ]:
L

[1, 2, 3, [4, 5]]

# *Mutability*

**Mutability refers to the ability to change or edit data in it's memory location**

In [ ]:
a = "HELLO"

In [ ]:
id(a)

135606217839152

In [ ]:
a = a+"world"

In [ ]:
a

'HELLOworld'

In [ ]:
id(a)

135606218038384

In [ ]:
T = (1,2,3)

In [ ]:
id(T)

135606219049792

In [ ]:
T = T +(5,6)

In [ ]:
T

(1, 2, 3, 5, 6)

In [ ]:
id(T)

135606219081744

**immutable_data_type = String , int , float , Bool , Complex , Tuplet**

In [ ]:
print(immutable_data_type)

NameError: name 'immutable_data_type' is not defined

**mutable_data_type = (List,Set,Dictionary)**

In [ ]:
print(mutable_data_type)

NameError: name 'mutable_data_type' is not defined

In [ ]:
L = [1,2,3]

In [ ]:
L

[1, 2, 3]

In [ ]:
id(L)

135606217734336

In [ ]:
L.append(4)

In [ ]:
L

[1, 2, 3, 4]

In [ ]:
id(L)

135606217734336

# **Side Effect Of Mutation**

*Like while working and we need to copy something if we do Aliasing then the previous data will be affected*

*agr kisi ko tmne apne data diya aur usne append yh aliasing kr diya toh dono data mutated ho jyga aur dono change hoga isse previous data change ho jyga aur dikkat hogi isse bachne k liye CLONING k use krte hai  *

In [ ]:
L =[1,2,3]

In [ ]:
L1 = L

In [ ]:
L1

[1, 2, 3]

In [ ]:
L

[1, 2, 3]

In [ ]:
id(L)

135606217684864

In [ ]:
id(L1)

135606217684864

In [ ]:
L1.append(4)

In [ ]:
L1

[1, 2, 3, 4]

In [ ]:
id(L1)

135606217684864

In [ ]:
L

[1, 2, 3, 4]

# *Cloning*

In [ ]:
L

[1, 2, 3, 4]

In [ ]:
L1 = L[:]

In [ ]:
id(L)

135606217684864

In [ ]:
id(L1)

135606217679744

In [ ]:
L1.append(5)

In [ ]:
L1

[1, 2, 3, 4, 5]

In [ ]:
L

[1, 2, 3, 4]

In [ ]:
a = (1,2,3,[4,5])

In [ ]:
a

(1, 2, 3, [4, 5])

In [ ]:
a[-1][-1]=500

*will it happen or not*

**Yes, it will happen b/s last cell is list an dit is mutable**

In [ ]:
a

(1, 2, 3, [4, 500])

In [ ]:
a = [1,2,3,(4,5)]

In [ ]:
a[-1][-1]=500

TypeError: 'tuple' object does not support item assignment

will it happen

no

In [ ]:
a

[1, 2, 3, (4, 5)]

In [ ]:
a = [1,2]

In [ ]:
b = [3,4]

In [ ]:
c = (a,b)

In [ ]:
c

([1, 2], [3, 4])

In [ ]:
id(a)

135606217976256

In [ ]:
id(b)

135606217882816

In [ ]:
id(c)

135606217220480

will this code work

In [ ]:
c[0][0]=100

yes it will b/s firt element is made of sets and set is matable

In [ ]:
c

([100, 2], [3, 4])

will the id of a,b,c will change ?

no

In [ ]:
id(a)

135606217976256

In [ ]:
id(b)

135606217882816

In [ ]:
id(c)

135606217220480

In [ ]:
c[0]=c[0]+[5,6]

TypeError: 'tuple' object does not support item assignment

In [ ]:
c

([100, 2], [3, 4])

In [ ]:
c = c +(5,6)

In [ ]:
c

([100, 2], [3, 4], 5, 6)

In [ ]:
a

[100, 2]

In [ ]:
a =a+[5,6]

In [ ]:
a

[100, 2, 5, 6]